Add docker compose setup

milafrerichs opened the pull request:
## What does this PR do?

It adds docker-compose version 2 setup to create a local environment for tock. 
Make sure to have the latest version of Docker and docker-compose installed.

All commands are explained in the Readme.
This is part of the Micropurchase Experiment from 18f.

### A few comments
I opted to use shell scripts to run specific commands. Almost everything can be used in a production environment.

adelevie commented:
Thanks @milafrerichs. We'll review this tomorrow. The docs look really good, by the way.

adelevie commented:
Hi @milafrerichs, 

I'm reviewing this on a fresh Ubuntu (15.10) box. A couple things (I think they're all pretty small and fixable).

 I got some errors about not having a `.env` file:

```
➜  tock git:(CivicVision-328-docker-compose-setup) docker-compose build tock
ERROR: Couldn't find env file: /home/adelevie/tock/.env
```

I didn't see anything in the README about having one. So I commented out `env_file: .env` in the `docker-compose.yml` file and moved forward:

`docker-compose build tock` seemed to have worked ([output](https://gist.github.com/adelevie/b45003051440f15b3dff)).

Things seemed to break down with the migration script:

```
➜  tock git:(CivicVision-328-docker-compose-setup) ✗ ./scripts/docker-migrate.sh 
Creating network "tock_default" with the default driver
Pulling db (postgres:latest)...
latest: Pulling from library/postgres
fdd5d7827f33: Already exists
a3ed95caeb02: Already exists
beb59dc2ad34: Already exists
f42a5322ef13: Already exists
f6719ae287c6: Already exists
0dc08677d778: Already exists
5f3b03c1dd66: Already exists
4d4c6707d860: Already exists
11f8efebbb9b: Already exists
edefda034373: Already exists
5fe4bba523f0: Already exists
a2d55eea3342: Already exists
Digest: sha256:bf1d7cbeb86b76f4c0991061515ebb13c6b57a9588a92f795eaebaede1c3657b
Status: Downloaded newer image for postgres:latest
Creating tock_db_1
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 119, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 176, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.4/site-packages/psycopg2/__init__.py", line 164, in connect
    conn = _connect(dsn, connection_factory=connection_factory, async=async)
psycopg2.OperationalError: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 19, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.4/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 93, in handle
    executor = MigrationExecutor(connection, self.migration_progress_callback)
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/executor.py", line 19, in __init__
    self.loader = MigrationLoader(self.connection)
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/loader.py", line 47, in __init__
    self.build_graph()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/loader.py", line 191, in build_graph
    self.applied_migrations = recorder.applied_migrations()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/recorder.py", line 59, in applied_migrations
    self.ensure_schema()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/recorder.py", line 49, in ensure_schema
    if self.Migration._meta.db_table in self.connection.introspection.table_names(self.connection.cursor()):
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 162, in cursor
    cursor = self.make_debug_cursor(self._cursor())
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 135, in _cursor
    self.ensure_connection()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python3.4/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 119, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 176, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.4/site-packages/psycopg2/__init__.py", line 164, in connect
    conn = _connect(dsn, connection_factory=connection_factory, async=async)
django.db.utils.OperationalError: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
```

Do you know what could be causing this?

adelevie commented on tock/tock/settings/dev.py:
```diff
@@ -10,9 +11,9 @@
-        'USER': 'tock',
-        'PASSWORD': 'tock',
-        'HOST': 'localhost',                      # Empty for localhost through domain sockets or           '127.0.0.1' for localhost through TCP.
+        'USER': os.environ.get('DATABASE_USER'),
```
@milafrerichs could the problem be a mistmatch between these env vars and the vars defined in docker-compose.yml?

```
  db:
    image: postgres
    environment:
      - POSTGRES_USER=tock
      - POSTGRES_PASSWORD=tock
      - POSTGRES_DB=tock
      - PGPASSWORD=tock
      - PGPUSER=tock
```

milafrerichs replied:
No. These are for the postgres image to set the default user and database. 
And for psql to do the same.

The postgres image will create a new user `tock` with the password `tock` and a new database `tock` for this user.

milafrerichs commented:
@adelevie 

Thanks for reviewing. 
In the auction description you added this bullet point:
```
Edit the .env file (read this page for more information on how docker-compose handles environment variables)
```

That's why I added the `.env` file in the `docker-compose.yml`
I added a sample.env file and updated the documentation to reflect the requirements of a `.env` file

batemapf commented:
@milafrerichs getting the same / similar error to [issue above](https://github.com/18F/tock/pull/333#issuecomment-198128389):

```
➜  tock git:(CivicVision-328-docker-compose-setup) ✗ scripts/docker-migrate.sh
Creating network "tock_default" with the default driver
Pulling db (postgres:latest)...
latest: Pulling from library/postgres
fdd5d7827f33: Already exists
a3ed95caeb02: Pull complete
beb59dc2ad34: Pull complete
f42a5322ef13: Pull complete
f6719ae287c6: Pull complete
0dc08677d778: Pull complete
5f3b03c1dd66: Pull complete
4d4c6707d860: Pull complete
11f8efebbb9b: Pull complete
edefda034373: Pull complete
5fe4bba523f0: Pull complete
a2d55eea3342: Pull complete
Digest: sha256:bf1d7cbeb86b76f4c0991061515ebb13c6b57a9588a92f795eaebaede1c3657b
Status: Downloaded newer image for postgres:latest
Creating tock_db_1
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 119, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 176, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.4/site-packages/psycopg2/__init__.py", line 164, in connect
    conn = _connect(dsn, connection_factory=connection_factory, async=async)
psycopg2.OperationalError: could not connect to server: Connection refused
	Is the server running on host "db" (172.18.0.2) and accepting
	TCP/IP connections on port 5432?


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 19, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.4/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 93, in handle
    executor = MigrationExecutor(connection, self.migration_progress_callback)
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/executor.py", line 19, in __init__
    self.loader = MigrationLoader(self.connection)
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/loader.py", line 47, in __init__
    self.build_graph()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/loader.py", line 191, in build_graph
    self.applied_migrations = recorder.applied_migrations()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/recorder.py", line 59, in applied_migrations
    self.ensure_schema()
  File "/usr/local/lib/python3.4/site-packages/django/db/migrations/recorder.py", line 49, in ensure_schema
    if self.Migration._meta.db_table in self.connection.introspection.table_names(self.connection.cursor()):
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 162, in cursor
    cursor = self.make_debug_cursor(self._cursor())
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 135, in _cursor
    self.ensure_connection()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python3.4/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 130, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/base/base.py", line 119, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.4/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 176, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.4/site-packages/psycopg2/__init__.py", line 164, in connect
    conn = _connect(dsn, connection_factory=connection_factory, async=async)
django.db.utils.OperationalError: could not connect to server: Connection refused
	Is the server running on host "db" (172.18.0.2) and accepting
	TCP/IP connections on port 5432?
```

milafrerichs commented:
Thanks. I will clarify in the readme. The db has to be up when you want to run the migrations.

So `docker-compose up` should be run first and afterwards you can run the migrations. 
Does that make sense or should I add more to it?

adelevie commented:
Hi @milafrerichs,

We tested this some more on two different machines (both OSX) and got the same results. The setup _mostly_ works (setting up the database*, inserting sample data, running the test suite**, and visiting `/employees` all work), but a few things are still lingering. Here's a [gist](https://gist.github.com/adelevie/9da653edf637e0321dda) of an entire terminal session of my results, but I'll summarize the issues below:

1) *`scripts/docker-migrate.sh` does not work on the first run. However on the 2nd run, it does work. We'd prefer if this worked the first time, but at a minimum, if it might need to be run twice, then that should be documented. Do you have any idea what could be causing this? (it could very well have to with our local docker (and docker-machine) installations)

2) While `/employees` and other pages are accessible (behind the IP you provided--also discoverable via `docker-machine ip`), `/` returns a `403 Forbidden` (we did state in the requirements that we'd need to be able to view the homepage of the site). I'm wondering if this is due to an error in the `docker-compose` setup or if it's a Tock application-level error. If the latter, what's the best way to peer into the Django logs? For what it's worth, `docker attach [container id]` showed no output.

3) **Running the test suite throws some errors. But the suite does run. If you can figure that out and fix it, great--but we will not hold it against you otherwise.

Anyways, we're really close to accepting this. If you can point out (with logs pasted in a gist maybe) that a) the setup works on your machine and/or b) that there is an issue with out setups that we can fix to make this all work, then we're likely to accept.

milafrerichs commented:
Hi @adelevie 

Thanks for looking into it.
I updated the Readme to clarify it a bit. To run the migrations an instance should be running otherwise it cannot connect to the database. (It has to be linked, that is done in the compose setup)

The tests should now only run with one error which can be ignored (if the test database already exists).
I could add to remove the test database if you want.

Regarding the issue that `/` returns a 404. I confirm that this is the case on my machine as well. But I thought this is an issue with your app.
The log (run with `docker-compose logs tock`) returns the following:
```
tock_1 | /usr/local/lib/python3.4/importlib/_bootstrap.py:321: RemovedInDjango19Warning: The utilities in django.db.models.loading are deprecated in favor of the new application loading system.
tock_1 |   return f(*args, **kwds)
tock_1 |
```
Not sure what is causing this.
If you want I can attach a log of my working setup. But would be best if will work on your machines.

adelevie commented:
Thanks @milafrerichs. The updated docs definitely help here. I was able to get `/employees` and tests working on the first run on a fresh Ubuntu box. 

To get the correct IP, I ran the following:

```
➜  tock git:(CivicVision-328-docker-compose-setup) docker ps                   
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                  NAMES
a5510c5470db        tock_tock           "python manage.py run"   24 minutes ago      Up 7 minutes        0.0.0.0:80->4001/tcp   tock_tock_1
81fb72e0c643        postgres            "/docker-entrypoint.s"   24 minutes ago      Up 7 minutes        5432/tcp               tock_db_1
➜  tock git:(CivicVision-328-docker-compose-setup) docker exec -it a5510c5470db bash
root@a5510c5470db:/tock# ip add | grep global
    inet 172.18.0.2/16 scope global eth0
``` 

Here, the IP is `172.18.0.2/16`. Adding port `4001` made this all work.

adelevie commented:
I'm trying to debug the `403 Forbidden` error. It seems that `docker-compose logs tock` is not adding new output when I make different requests to the local site. Are there alternative ways to view the Django logs?

milafrerichs commented:
The `docker-compose logs tock` is not adding output because you did not specify a logging scheme in Django.
For the production environment you're doing it [here](https://github.com/18F/tock/blob/master/tock/tock/settings/production.py#L22)

The output I get when I add the logging setup:

```
tock_1 | [19/Mar/2016 16:43:44] "GET / HTTP/1.1" 403 22
tock_1 | Forbidden (Permission denied): /
```

milafrerichs commented:
And for the IP:

I see you're using `docker-machine` with `docker-machine ls` you're getting the IP of your box.
